platform/api/hetzner/api.go

```diff
@@ -27,26 +41,341 @@ type Options struct {
 
 // API is a wrapper around Hetzner instance API.
 type API struct {
-	opts   *Options
-	client *hcloud.Client
+	client       *hcloud.Client
+	imagesclient *hcloudimages.Client
+
+	serverType *hcloud.ServerType
+	image      *hcloud.Image
+	location   *hcloud.Location
 }
 
 type Server struct{ *hcloud.Server }
 
+type SSHKey struct{ *hcloud.SSHKey }
+
+type Network struct{ *hcloud.Network }
+
 // New returns a Hetzner API instance.
 func New(opts *Options) (*API, error) {
-	return nil, nil
+	client := hcloud.NewClient(
+		hcloud.WithToken(opts.Token),
+		hcloud.WithApplication("flatcar-mantle", version.Version),
+		hcloud.WithPollBackoffFunc(hcloud.ExponentialBackoff(2, 500*time.Millisecond)),
+	)
+
+	imagesclient := hcloudimages.NewClient(client)
+
+	ctx := context.Background()
+
+	serverType, _, err := client.ServerType.Get(ctx, opts.ServerType)
+	if err != nil {
+		return nil, fmt.Errorf("verifying server type: %w", err)
+	}
+
+	var image *hcloud.Image
+	if opts.Image != "" {
+		image, _, err = client.Image.GetForArchitecture(ctx, opts.Image, serverType.Architecture)
+		if err != nil {
+			return nil, fmt.Errorf("verifying image: %w", err)
+		}
+	}
+
+	location, _, err := client.Location.Get(ctx, opts.Location)
+	if err != nil {
+		return nil, fmt.Errorf("verifying location: %w", err)
+	}
+
+	return &API{
+		client:       client,
+		imagesclient: imagesclient,
+
+		serverType: serverType,
+		image:      image,
+		location:   location,
+	}, nil
 }
 
 // CreateServer using the Hetzner API.
-func (a *API) CreateServer(ctx context.Context, name, userdata string) (*Server, error) {
-	return nil, nil
+func (a *API) CreateServer(ctx context.Context, name, userdata string, sshKey *SSHKey, network *Network) (*Server, error) {
+	opts := hcloud.ServerCreateOpts{
+		Name:       name,
+		ServerType: a.serverType,
+		Image:      a.image,
+		Location:   a.location,
+		UserData:   userdata,
+	}
+	if sshKey != nil {
+		opts.SSHKeys = []*hcloud.SSHKey{sshKey.SSHKey}
+	}
+	if network != nil {
+		opts.Networks = []*hcloud.Network{network.Network}
+	}
+
+	result, _, err := a.client.Server.Create(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to request new server: %w", err)
+	}
+
+	err = a.client.Action.WaitFor(ctx, append(result.NextActions, result.Action)...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new server: %w", err)
+	}
+
+	return &Server{result.Server}, nil
```

Comment:
Mh, I could not reproduce this locally. I changed the code to do a single reload after all actions have settled if a network is set and the server does not contain the IP.

The suggested code in particular had a tight loop that was calling the API. Something we would very much like to avoid for the (user) risk of exhausting the rate limit and the (server) risk of DoS.